### ISSUE TITLE
Is the ability of an InProofV2 in the Zero to Monero document overstated?

### ISSUE THREAD

**Issue by ChrisCharrison** [monero-project/monero#7353]
Here is the document. See the bottom of page 79 onwards for InProofV2.
[https://web.getmonero.org/library/Zero-to-Monero-2-0-0.pdf](url)

The document correctly outlines how Outproofs can be 'faked' at the bottom of page 78. To bring others up to speed, one can take a stealth address:

`P = H(rA) + B`

and rearrange like:

`P - H(rA)G = B`
[H is a scalar derivation, G is the ec-basepoint, r is the tx private key, `A(,a)` and `B(,b) `are the public (private) view and spend keys (of the owner) respectively]

and plug some arbitrary A' into the left hand side and derive a corresponding B', such that A' and B' still satisfy the original stealth address P. One then presents `(A', B', r)` to the verifier who then naively 'confirms' that the Monero address `(A,B)` matches the output. For completeness it's worth saying that despite procuring such an `(A', B')` the 'faker' does not have access to the private counterpart of B' (=b'), and therefore cannot spend the output. 

The document then goes on to specify another proof called 'InProof V2' which is meant to be a proof of the other side of the sender-receiver shared secret `rA =aR`; that is, the prover is supposed to 'prove knowledge' of the private view key for the Monero address that owns a particular output. However, again we are permitted to choose some arbitrary A' (and B') to satisfy the stealth address in question. 

The Z2M document makes the following claim however:

> Prove ‘full’ ownership with the one-time address key:
...an InProof shows a one-time address was constructed with a specific address (except with
negligible probability)

and then later details the actual way of fully proving ownership:
> Proving ownership, once an InProof is complete, is as simple as signing a message with the spend
key.

My contention is that the statement 

> an InProof shows a one-time address was constructed with a specific address (except with
negligible probability)

is incorrect, as any A',B' that satisfy P are equal candidates at the InProof stage; an InProof without a corresponding spend-key signature means nothing. It might be worth migrating the spend-key signature step into the actual proof step-by-step guide, beings as it's so critical.

All the best,

Chris

**Comment by UkoeHB** [comment]
Chris,

I see what you're saying... without a proof on `K^s`, it could be any random curve point. In other words you could misrepresent ownership of a single output with a fake InProofV2 (although the address couldn't own any other outputs except with negligible probability).

Even with a proof/signature on `K^s`, it is possible for the sets of outputs 'owned' by two addresses to overlap in one output (as provable using InProofV2s, which at least lock down the addressing scheme), if the second address is derived from `k^o` in that output. In other words, randomly generate `k'^v`, and derive `k'^s` from `k'^s = k^o - H(k'^v*rG, t)`. This key pair `k'^v, k'^s` can be used to receive new outputs. I don't think there is a way to detect this overlap trick.

**Comment by ChrisCharrison** [comment]
Philosophically you are only proving the spendkey of the destination XMR address by doing the final signature, as there are many XMR addresses that contain the same spend key but different view keys. However @UkoeHB one thing you're missing that guards against overlap tricks is the following:

`P - H(rA)G = B`

You may choose an arbitrary `a'`, compute `a'G=A'`, and find the corresponding `B'` by plugging into the LHS. However in doing so you are unable to compute `b'` for any given `a'` that you feed in. This means that if you are signing some message with any b' that satisfies `P`, you are indeed the owner of the output (except with negligible probability). Just for completeness, it's worth saying that starting with a `b'`, deriving `b'G=B'` and getting the corresponding `A'` (even if you know `r`) because this constitutes performing `H^-1` (reversing Keccak) which is computationally infeasible.

**Comment by UkoeHB** [comment]
@ChrisCharrison the overlap trick only works for outputs you actually own, because then you know the private keys in `p - H(rA) = b` and can compute `p - H(rA') = b'` then use `a', b'` as a new view/spend pair.

**Comment by ChrisCharrison** [comment]
That's absolutely right @UkoeHB . For those reading the full proof of why this works is:

`P - H(rA)G = pG + H(rA)G`   writing both terms as scalar multiplication
`pG - H(rA)G = (p -H(rA))G`  because point addition is additive 
`(p-H(rA))G = bG`                  as before
`=> p-H(rA) = b`

You can trivially find multiple pairs of `(a', b')` that satisfy P if you own the output (and therefore know what the output private key `p` is). We actually sign outputs with the _output_ private key and not the spend key when we sign a transaction. You can do an valid inproof an then sign afterwards with any spendkey that fits the equation for P, but that only proves that you able to spend the output because you have by proxy proven that you know the output private key. By no means are you proving that your output belongs to a certain XMR address (A,B) just by signing a message with a spendkey. @UkoeHB surely this means that we simply never know who an output truly belongs to? Because once you can compute multiple (a',b') and that allow you to reach the same output private key, you could claim plausible deniability over which original (a,b) you used to achieve that output private key in the first place?

**Comment by UkoeHB** [comment]
@ChrisCharrison it is not quite that severe. If your address `(A', B')` owns more than one output (as shown with InProofs), then any outputs after the first must have been intentionally sent to that address by the sender.

The importance of InProofs is, given _some publicly known address_, you can demonstrate the full set of outputs owned by that address. After all, you can have any number of private addresses which own any number of outputs.

This overlap trick is only useful in niche cases where you want to demonstrate ownership of an output without revealing which address the sender used to construct it with.

**Comment by ChrisCharrison** [comment]
That is what I said/meant in essence. 
The main issue is just what the documentation of inproof says it can achieve (prove a specific address owns an output, which it can't). This could have implications for accounting though as well, but I know Monero is mainly about being as untraceable as possible. You can only prove that you _can_ spend an output though, that's my point. 

Even for a set of multiple inproofs, one could be a fake. You could do get your 'fake' (a',b') for the first output, and then continue to use that address afterwards for transactions legitimately sent to you by others. So you cannot fully demonstrate the full set of outputs owned by an XMR address. I just think it's worth people being aware of that/the documentation reflecting this stuff.

Thanks for the discussion.

**Comment by ChrisCharrison** [comment]
Yes agreed. That is what I meant to get across - there is no unique private spendkey that solely owns the right to spend a given output, and one person can hold many of those keys. Most people just think of multiple addresses being able to spend the same output because they have different view keys but share the same spend key. It's a subtle but interesting feature of the Monero cryptography. To comment on your final sentence (and why I asked about all this in the first place) - we run a less private fork of Monero and for accounting/regulatory purposes it's much desirable to know exact addresses to know the exact flow of tokens. Thanks for the discussion @UkoeHB.

**Comment by UkoeHB** [comment]
I looked at this again. `InProofV2`s are checking amount commitments the same way `OutProofV2`s are. If an amount commitment is successfully checked, then its output must be spendable by the `Kv, Ks` pair used to make the `InProofV2` (if it is spendable at all under the standard address rules). Therefore, it shouldn't be necessary to sign on `Ks` in the `InProofV2`. If you want to prove that you can spend all the outputs mentioned in inproofs, then you need to sign on all your `Ks`s.
